File: archive/audit_feed_exploration/bridge_audit_feed.py
"""
ATI Audit Feed Bridge - All Active AMRs to Twinzo Old Plant

This bridge uses the ATI audit feed credentials to monitor ALL active AMRs
and stream their positions to the appropriate Twinzo devices in Old Plant (Sector 2).

Features:
- Monitors ati_fm/sherpa/status for ALL AMRs
- Maps ATI sherpa_name to Twinzo device logins
- Streams to Old Plant (Sector 2) only
- Handles 7 active AMRs from multiple fleets

Usage:
    python -X utf8 src/bridge/bridge_audit_feed.py
"""

import os
import ssl
import time
import json
import requests
from dotenv import load_dotenv
import paho.mqtt.client as mqtt

# Load environment variables
load_dotenv()

# ATI Audit Feed Configuration
ATI_HOST = os.getenv("AUDIT_MQTT_HOST", "tvs-dev.ifactory.ai")
ATI_PORT = int(os.getenv("AUDIT_MQTT_PORT", "8883"))
ATI_USERNAME = os.getenv("AUDIT_USERNAME", "tvs-audit-user")
ATI_PASSWORD = os.getenv("AUDIT_PASSWORD", "TVSAudit@2025")
ATI_CLIENT_ID = "audit-bridge-old-plant"
ATI_TOPIC = "ati_fm/sherpa/status"

# Twinzo API Configuration
TWINZO_CLIENT = os.getenv("TWINZO_CLIENT", "TVSMotor")
TWINZO_PASSWORD = os.getenv("TWINZO_PASSWORD")
TWINZO_API_KEY = os.getenv("TWINZO_API_KEY")
TWINZO_AUTH_URL = "https://api.platform.twinzo.com/v3/authorization/authenticate"
TWINZO_LOC_URL = "https://api.platform.twinzo.com/v3/localization"

# Old Plant Configuration
OLD_PLANT_SECTOR = 2
OLD_PLANT_BRANCH = "40557468-2d57-4a3d-9a5e-3eede177daf5"

# Device Mapping: ATI sherpa_name -> Twinzo login
DEVICE_MAP = {
    "tug-55-tvsmotor-hosur-09": "tug-55-hosur-09",
    "tug-39-tvsmotor-hosur-07": "tug-39-hosur-07",
    "tug-133": "tug-133",
    "tug-140": "tug-140",
    "tug-78": "tug-78",
    "tug-24-tvsmotor-hosur-05": "tug-24-hosur-05",
    "tug-11": "tug-11"
}

# Coordinate transformation (if needed - adjust based on actual ATI coordinates)
AFFINE_A = float(os.getenv("AFFINE_A", "1.0"))
AFFINE_B = float(os.getenv("AFFINE_B", "0.0"))
AFFINE_C = float(os.getenv("AFFINE_C", "0.0"))
AFFINE_D = float(os.getenv("AFFINE_D", "1.0"))
AFFINE_TX = float(os.getenv("AFFINE_TX", "0.0"))
AFFINE_TY = float(os.getenv("AFFINE_TY", "0.0"))

# OAuth token cache
oauth_cache = {
    "tokens": {},
    "last_cleanup": time.time()
}

# Statistics
stats = {
    "messages_received": 0,
    "messages_sent": 0,
    "last_update": {},
    "errors": 0
}

# Requests session for connection pooling
session = requests.Session()


def transform_xy(x, y):
    """Apply affine transformation to coordinates"""
    return AFFINE_A*x + AFFINE_B*y + AFFINE_TX, AFFINE_C*x + AFFINE_D*y + AFFINE_TY


def authenticate_device(device_login):
    """Authenticate Twinzo device"""
    try:
        r = session.post(TWINZO_AUTH_URL, json={
            "client": TWINZO_CLIENT,
            "login": device_login,
            "password": TWINZO_PASSWORD
        }, timeout=10)

        if r.status_code == 200:
            d = r.json()
            oauth_cache["tokens"][device_login] = {
                "token": d["Token"],
                "client": d["Client"],
                "branch": d["Branch"],
                "expires": d["Expiration"]
            }
            print(f"OK OAuth successful for {device_login}")
            return oauth_cache["tokens"][device_login]
        else:
            print(f"FAIL OAuth failed for {device_login}: {r.status_code}")
            return None
    except Exception as e:
        print(f"FAIL OAuth error for {device_login}: {e}")
        return None


def get_device_credentials(device_login):
    """Get cached or fresh credentials"""
    now = time.time()

    # Cleanup expired tokens
    if now - oauth_cache["last_cleanup"] > 600:
        expired = [d for d, c in oauth_cache["tokens"].items() if c["expires"] < now * 1000]
        for d in expired:
            del oauth_cache["tokens"][d]
        oauth_cache["last_cleanup"] = now

    # Check if we have valid token
    if device_login in oauth_cache["tokens"]:
        creds = oauth_cache["tokens"][device_login]
        if creds["expires"] > (now + 60) * 1000:
            return creds

    # Get fresh token
    return authenticate_device(device_login)


def send_to_twinzo(device_login, x, y, heading):
    """Send position data to Twinzo"""
    try:
        creds = get_device_credentials(device_login)
        if not creds:
            return False

        # Prepare localization data
        data = {
            "SectorId": OLD_PLANT_SECTOR,
            "X": x,
            "Y": y,
            "Heading": heading
        }

        # Send to Twinzo
        headers = {
            "Authorization": f"Bearer {creds['token']}",
            "apikey": TWINZO_API_KEY
        }

        r = session.post(TWINZO_LOC_URL, json=data, headers=headers, timeout=5)

        if r.status_code in [200, 201, 204]:
            stats["messages_sent"] += 1
            return True
        else:
            print(f"FAIL Twinzo API error for {device_login}: {r.status_code}")
            stats["errors"] += 1
            return False

    except Exception as e:
        print(f"FAIL Error sending to Twinzo for {device_login}: {e}")
        stats["errors"] += 1
        return False


def on_connect(client, userdata, flags, rc, properties=None):
    """Callback when connected to ATI broker"""
    if rc == 0:
        print("OK Connected to ATI audit feed")
        client.subscribe(ATI_TOPIC, qos=1)
        print(f"OK Subscribed to {ATI_TOPIC}")
    else:
        print(f"FAIL Connection failed with code {rc}")


def on_message(client, userdata, message):
    """Callback when MQTT message received"""
    try:
        stats["messages_received"] += 1

        # Parse the message
        payload = json.loads(message.payload.decode('utf-8'))

        sherpa_name = payload.get("sherpa_name")
        mode = payload.get("mode")
        pose = payload.get("pose")
        battery = payload.get("battery_status", 0)

        # Skip if not in our device map
        if sherpa_name not in DEVICE_MAP:
            return

        # Skip if disconnected or no position data
        if mode != "fleet" or not pose or len(pose) < 3:
            return

        # Get Twinzo device login
        device_login = DEVICE_MAP[sherpa_name]

        # Extract position
        x_raw, y_raw, heading = pose[0], pose[1], pose[2]

        # Apply coordinate transformation
        x, y = transform_xy(x_raw, y_raw)

        # Send to Twinzo
        success = send_to_twinzo(device_login, x, y, heading)

        # Update stats
        now = time.time()
        stats["last_update"][sherpa_name] = now

        # Log periodically
        if stats["messages_received"] % 50 == 0:
            active_count = len([t for t in stats["last_update"].values() if now - t < 30])
            print(f"\nStats: Received={stats['messages_received']}, "
                  f"Sent={stats['messages_sent']}, Active={active_count}, "
                  f"Errors={stats['errors']}")
            print(f"Recent updates: {list(stats['last_update'].keys())}\n")

    except json.JSONDecodeError:
        print(f"FAIL Invalid JSON in message")
        stats["errors"] += 1
    except Exception as e:
        print(f"FAIL Error processing message: {e}")
        stats["errors"] += 1


def main():
    """Main bridge function"""
    print("=" * 70)
    print("ATI Audit Feed Bridge - Old Plant (Sector 2)")
    print("=" * 70)
    print(f"ATI Broker: {ATI_HOST}:{ATI_PORT}")
    print(f"ATI Username: {ATI_USERNAME}")
    print(f"Client ID: {ATI_CLIENT_ID}")
    print(f"Topic: {ATI_TOPIC}")
    print(f"Old Plant Sector: {OLD_PLANT_SECTOR}")
    print(f"Old Plant Branch: {OLD_PLANT_BRANCH}")
    print(f"Active AMRs: {len(DEVICE_MAP)}")
    for ati_name, twinzo_login in DEVICE_MAP.items():
        print(f"  {ati_name} -> {twinzo_login}")
    print("=" * 70)

    if not ATI_USERNAME or not ATI_PASSWORD:
        print("\nFAIL ATI audit credentials not configured!")
        print("Set AUDIT_USERNAME and AUDIT_PASSWORD environment variables")
        return

    if not TWINZO_PASSWORD or not TWINZO_API_KEY:
        print("\nFAIL Twinzo credentials not configured!")
        print("Set TWINZO_PASSWORD and TWINZO_API_KEY environment variables")
        return

    # Create MQTT client
    client = mqtt.Client(
        mqtt.CallbackAPIVersion.VERSION2,
        client_id=ATI_CLIENT_ID,
        protocol=mqtt.MQTTv5
    )
    client.username_pw_set(ATI_USERNAME, ATI_PASSWORD)

    # TLS configuration
    client.tls_set(cert_reqs=ssl.CERT_NONE)
    client.tls_insecure_set(True)

    # Set callbacks
    client.on_connect = on_connect
    client.on_message = on_message

    print(f"\nConnecting to ATI audit feed...")
    print(f"Protocol: MQTT v5, QoS: 1")

    try:
        client.connect(ATI_HOST, ATI_PORT, keepalive=60)
        print("OK Bridge ready - streaming 7 AMRs to Old Plant\n")
        client.loop_forever()
    except KeyboardInterrupt:
        print("\nShutting down bridge...")
        print(f"Final stats: Received={stats['messages_received']}, "
              f"Sent={stats['messages_sent']}, Errors={stats['errors']}")
        client.disconnect()
        print("OK Bridge stopped")
    except Exception as e:
        print(f"FAIL Bridge error: {e}")


if __name__ == "__main__":
    main()

File: src/bridge/bridge_audit_feed.js
/**
 * ATI Audit Feed Bridge - Old Plant AMRs to Twinzo
 *
 * This bridge uses the ATI audit feed credentials to monitor active AMRs
 * and stream their positions to Twinzo Old Plant (Sector 2).
 *
 * Features:
 * - Monitors ati_fm/sherpa/status for all AMRs
 * - Maps ATI sherpa_name to Twinzo device logins
 * - Streams to Old Plant (Sector 2) only
 * - Currently handling 3 AMRs: tug-55, tug-39, tug-133
 * - Uses affine transformation for coordinate mapping
 * - Logs all data to SQLite database (logs/ati_data.db)
 *
 * IMPORTANT - ATI Timestamp Format:
 * - ATI includes a "timestamp" field in ISO 8601 format with 'Z' suffix
 * - Example: "timestamp": "2025-12-01T13:21:52Z"
 * - WARNING: Despite the 'Z' (UTC indicator), timestamps are actually in IST (UTC+5:30)
 * - This is a data formatting issue on ATI's side
 * - Bridge uses Date.now() for Twinzo API (real-time when received)
 *
 * Usage:
 *     node src/bridge/bridge_audit_feed.js
 */

import mqtt from 'mqtt';
import 'dotenv/config';
import fetch from 'node-fetch';
import { logATIMessage } from '../common/database.js';

// ATI Audit Feed Configuration
const ATI_HOST = process.env.AUDIT_MQTT_HOST || 'tvs-dev.ifactory.ai';
const ATI_PORT = parseInt(process.env.AUDIT_MQTT_PORT || '8883', 10);
const ATI_USERNAME = process.env.AUDIT_USERNAME || 'tvs-audit-user';
const ATI_PASSWORD = process.env.AUDIT_PASSWORD || 'TVSAudit@2025';
// Try exact match first - audit credentials require client ID = username
const ATI_CLIENT_ID = ATI_USERNAME;
// Subscribe to ATI topics (same format as Pavan's working code)
const SUBSCRIPTION_TOPICS = {
    'ati_fm/#': {qos: 1},
    'fleet/trips/info': {qos: 1}
};

// Twinzo API Configuration
const TWINZO_CLIENT = process.env.TWINZO_CLIENT || 'TVSMotor';
const TWINZO_PASSWORD = process.env.TWINZO_PASSWORD;
const TWINZO_API_KEY = process.env.TWINZO_API_KEY;
const TWINZO_AUTH_URL = 'https://api.platform.twinzo.com/v3/authorization/authenticate';
const TWINZO_LOC_URL = 'https://api.platform.twinzo.com/v3/localization';

// Plant Configuration
const OLD_PLANT_SECTOR = 2;
const OLD_PLANT_BRANCH = '40557468-2d57-4a3d-9a5e-3eede177daf5';
const HITECH_PLANT_SECTOR = 1;
const HITECH_PLANT_BRANCH = 'dcac4881-05ab-4f29-b0df-79c40df9c9c2';

// Device Mapping: ATI sherpa_name -> Twinzo login (ONLY OLD PLANT)
const DEVICE_MAP = {
    'tug-55-tvsmotor-hosur-09': 'tug-55-hosur-09',
    'tug-39-tvsmotor-hosur-07': 'tug-39-hosur-07',
    'tug-133': 'tug-133'
    // Disabled: tug-140, tug-78, tug-24, tug-11 (not in Old Plant)
};

// Device-to-Sector Mapping: Which sector each device belongs to
const DEVICE_SECTOR_MAP = {
    'tug-55-hosur-09': OLD_PLANT_SECTOR,
    'tug-39-hosur-07': OLD_PLANT_SECTOR,
    'tug-133': OLD_PLANT_SECTOR
};

// Coordinate transformation: Scale ATI meters (0-120 range) to Twinzo units (100k range)
// ATI coordinates are in meters, Twinzo uses ~100k scale
// Transform: Scale by 1000, then offset to place in Old Plant map area
const AFFINE_A = parseFloat(process.env.AFFINE_A || '1000.0');  // X scale
const AFFINE_B = parseFloat(process.env.AFFINE_B || '0.0');     // X rotation
const AFFINE_C = parseFloat(process.env.AFFINE_C || '0.0');     // Y rotation
const AFFINE_D = parseFloat(process.env.AFFINE_D || '1000.0');  // Y scale
const AFFINE_TX = parseFloat(process.env.AFFINE_TX || '100000.0');  // X offset (center at 100k)
const AFFINE_TY = parseFloat(process.env.AFFINE_TY || '100000.0');  // Y offset (center at 100k)

// OAuth token cache
const oauthCache = {};

// Statistics
const stats = {
    messagesReceived: 0,
    messagesTotal: 0,  // All messages including non-sherpa
    messagesSent: 0,
    errors: 0,
    lastUpdate: {}
};

function transformXY(x, y) {
    return [
        AFFINE_A * x + AFFINE_B * y + AFFINE_TX,
        AFFINE_C * x + AFFINE_D * y + AFFINE_TY
    ];
}

async function authenticateDevice(deviceLogin) {
    try {
        const response = await fetch(TWINZO_AUTH_URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                client: TWINZO_CLIENT,
                login: deviceLogin,
                password: TWINZO_PASSWORD
            }),
            timeout: 10000
        });

        if (response.ok) {
            const data = await response.json();
            oauthCache[deviceLogin] = {
                token: data.Token,
                client: data.Client,
                branch: data.Branch,
                expires: data.Expiration
            };
            console.log(`OK OAuth successful for ${deviceLogin}`);
            return oauthCache[deviceLogin];
        } else {
            console.log(`FAIL OAuth failed for ${deviceLogin}: ${response.status}`);
            return null;
        }
    } catch (error) {
        console.log(`FAIL OAuth error for ${deviceLogin}: ${error.message}`);
        return null;
    }
}

async function getDeviceCredentials(deviceLogin) {
    const now = Date.now();

    // Check if we have valid token
    if (oauthCache[deviceLogin]) {
        const creds = oauthCache[deviceLogin];
        if (creds.expires > now + 60000) {
            return creds;
        }
    }

    // Get fresh token
    return await authenticateDevice(deviceLogin);
}

async function sendToTwinzo(deviceLogin, x, y, heading, battery = 0) {
    try {
        const creds = await getDeviceCredentials(deviceLogin);
        if (!creds) {
            return { success: false, error: 'Authentication failed' };
        }

        // Validate coordinates
        if (!isFinite(x) || !isFinite(y)) {
            console.log(`FAIL Invalid coordinates for ${deviceLogin}: X=${x}, Y=${y}`);
            return { success: false, error: `Invalid coordinates: X=${x}, Y=${y}` };
        }

        // Get device-specific sector (default to Old Plant if not configured)
        const sectorId = DEVICE_SECTOR_MAP[deviceLogin] || OLD_PLANT_SECTOR;

        // Prepare localization data (MUST be an array, match Python bridge format)
        const data = [{
            Timestamp: Date.now(),
            SectorId: sectorId,
            X: x,
            Y: y,
            Z: 0,
            Interval: 100,
            Battery: battery,
            IsMoving: true,
            LocalizationAreas: [],
            NoGoAreas: []
        }];

        // Send to Twinzo (use exact header format from Python bridge)
        const response = await fetch(TWINZO_LOC_URL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Client': creds.client,
                'Branch': creds.branch,
                'Token': creds.token,
                'Api-Key': TWINZO_API_KEY
            },
            body: JSON.stringify(data),
            timeout: 5000
        });

        if (response.ok || response.status === 204) {
            stats.messagesSent++;
            return { success: true, response: `HTTP ${response.status}` };
        } else {
            const errorText = await response.text();
            console.log(`FAIL Twinzo API error for ${deviceLogin}: ${response.status}`);
            console.log(`  Sector: ${sectorId}, X: ${x.toFixed(2)}, Y: ${y.toFixed(2)}, Battery: ${battery}`);
            console.log(`  Response: ${errorText.substring(0, 200)}`);
            stats.errors++;
            return { success: false, error: `HTTP ${response.status}: ${errorText.substring(0, 200)}` };
        }
    } catch (error) {
        console.log(`FAIL Error sending to Twinzo for ${deviceLogin}: ${error.message}`);
        stats.errors++;
        return { success: false, error: error.message };
    }
}

// Connection options
const connectionOptions = {
    protocol: 'mqtts',
    host: ATI_HOST,
    port: ATI_PORT,
    clientId: ATI_CLIENT_ID,
    username: ATI_USERNAME,
    password: ATI_PASSWORD,
    protocolVersion: 5,
    clean: true,  // Use clean session to avoid conflicts
    reconnectPeriod: 5000,
    connectTimeout: 30000,
    rejectUnauthorized: false,
    properties: {
        sessionExpiryInterval: 0  // No persistent session
    }
};

console.log('======================================================================');
console.log('ATI Audit Feed Bridge - Old Plant (Sector 2)');
console.log('======================================================================');
console.log(`ATI Broker: ${ATI_HOST}:${ATI_PORT}`);
console.log(`ATI Username: ${ATI_USERNAME}`);
console.log(`Client ID: ${ATI_CLIENT_ID}`);
console.log(`Topics: ${Object.keys(SUBSCRIPTION_TOPICS).join(', ')}`);
console.log(`Old Plant - Sector: ${OLD_PLANT_SECTOR}, Branch: ${OLD_PLANT_BRANCH}`);
console.log(`HiTech Plant - Sector: ${HITECH_PLANT_SECTOR}, Branch: ${HITECH_PLANT_BRANCH}`);
console.log(`Coordinate Transform: A=${AFFINE_A}, B=${AFFINE_B}, C=${AFFINE_C}, D=${AFFINE_D}, TX=${AFFINE_TX}, TY=${AFFINE_TY}`);
console.log(`Active AMRs: ${Object.keys(DEVICE_MAP).length}`);
for (const [atiName, twinzoLogin] of Object.entries(DEVICE_MAP)) {
    const sector = DEVICE_SECTOR_MAP[twinzoLogin] || OLD_PLANT_SECTOR;
    const plantName = sector === HITECH_PLANT_SECTOR ? 'HiTech' : 'Old Plant';
    console.log(`  ${atiName} -> ${twinzoLogin} (${plantName}, Sector ${sector})`);
}
console.log('======================================================================\\n');

const client = mqtt.connect(connectionOptions);

client.on('connect', () => {
    console.log('OK Connected to ATI audit feed');
    client.subscribe(SUBSCRIPTION_TOPICS, (err, granted) => {
        if (err) {
            console.error('FAIL Subscription error:', err);
            return;
        }
        console.log('OK Subscribed to:');
        for (const sub of granted) {
            console.log(`  ${sub.topic}`);
        }
        console.log('OK Bridge ready - streaming 3 AMRs to Old Plant (Sector 2)\\n');
    });
});

client.on('message', async (topic, messageBuffer) => {
    try {
        stats.messagesTotal++;

        // Log every 20 messages to show activity
        if (stats.messagesTotal % 20 === 0) {
            console.log(`DEBUG: Total=${stats.messagesTotal}, Processed=${stats.messagesReceived}, Sent=${stats.messagesSent}, topic=${topic}`);
        }

        // Only process sherpa status messages
        if (topic !== 'ati_fm/sherpa/status') {
            return;
        }

        stats.messagesReceived++;

        const payload = JSON.parse(messageBuffer.toString());
        const sherpaName = payload.sherpa_name;
        const mode = payload.mode;
        const pose = payload.pose;
        const battery = payload.battery_status || 0;

        // Skip if not in our device map
        if (!DEVICE_MAP[sherpaName]) {
            return;
        }

        // Skip if disconnected or no position data
        if (mode !== 'fleet' || !pose || pose.length < 3) {
            return;
        }

        // Get Twinzo device login
        const deviceLogin = DEVICE_MAP[sherpaName];

        // Extract position
        const [xRaw, yRaw, heading] = pose;

        // Apply coordinate transformation
        const [x, y] = transformXY(xRaw, yRaw);

        // Log detailed data for every message
        console.log(`\n${'='.repeat(70)}`);
        console.log(`AMR: ${sherpaName} -> ${deviceLogin}`);
        console.log(`Time: ${new Date().toISOString()}`);
        console.log(`ATI Raw Coordinates: X=${xRaw.toFixed(2)}m, Y=${yRaw.toFixed(2)}m, Heading=${heading.toFixed(3)}rad`);
        console.log(`Twinzo Transformed: X=${x.toFixed(0)}, Y=${y.toFixed(0)}`);
        console.log(`Battery: ${battery}%, Mode: ${mode}`);
        console.log(`${'='.repeat(70)}`);

        // Send to Twinzo
        const result = await sendToTwinzo(deviceLogin, x, y, heading, battery);

        // Log to database
        try {
            logATIMessage({
                device_name: sherpaName,
                ati_x: xRaw,
                ati_y: yRaw,
                ati_heading: heading,
                twinzo_x: x,
                twinzo_y: y,
                twinzo_heading: heading,
                battery_status: battery,
                mode: mode,
                posted_to_api: result.success,
                api_response: result.response || null,
                error: result.error || null
            });
        } catch (dbError) {
            console.log(`WARN Database logging failed: ${dbError.message}`);
        }

        // Update stats
        stats.lastUpdate[sherpaName] = Date.now();

        // Log periodically
        if (stats.messagesReceived % 50 === 0) {
            const now = Date.now();
            const activeCount = Object.values(stats.lastUpdate).filter(t => now - t < 30000).length;
            console.log(`\\nStats: Received=${stats.messagesReceived}, ` +
                       `Sent=${stats.messagesSent}, Active=${activeCount}, ` +
                       `Errors=${stats.errors}`);
            console.log(`Recent updates: ${Object.keys(stats.lastUpdate).join(', ')}\\n`);
        }

    } catch (error) {
        console.error('FAIL Error processing message:', error.message);
        stats.errors++;
    }
});

client.on('error', (error) => {
    console.error('FAIL MQTT Error:', error.message);
});

client.on('reconnect', () => {
    console.log('Reconnecting to ATI audit feed...');
});

client.on('offline', () => {
    console.log('Client offline');
});

client.on('close', () => {
    console.log('Connection closed');
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\\n\\nShutting down bridge...');
    console.log(`Final stats: Received=${stats.messagesReceived}, ` +
               `Sent=${stats.messagesSent}, Errors=${stats.errors}`);
    client.end(false, () => {
        console.log('OK Bridge stopped');
        process.exit(0);
    });
});

process.on('SIGTERM', () => {
    console.log('\\n\\nShutting down bridge...');
    console.log(`Final stats: Received=${stats.messagesReceived}, ` +
               `Sent=${stats.messagesSent}, Errors=${stats.errors}`);
    client.end(false, () => {
        console.log('OK Bridge stopped');
        process.exit(0);
    });
});
